#357 - Syncing filter value with query param.

RubensKj opened the pull request:
Issue: #357

What I did was sync the filter value with the query param, so that's easier to share the extension and etc.. And it attend what the issue ask for.

ia3andy commented:
Hey @RubensKj, don't you think we should sync all fields so that the url is like the "share" feature?

I guess I would keep cleaning the url on load as we currently do, then rewrite it with the encoded project and search.. wdyt?

RubensKj commented:
> Hey @RubensKj, don't you think we should sync all fields so that the url is like the "share" feature?
> 
> I guess I would keep cleaning the url on load as we currently do, then rewrite it with the encoded project and search.. wdyt?

I like this idea of being all fields sync, but when implementing this, do you think we should remove the share button? because it will not be used  anymore (Instead we copy the url) .

ia3andy commented:
I will give it a try today :)

ia3andy commented on src/main/frontend/src/code-quarkus/api/quarkus-project-utils.ts:
```diff
@@ -94,6 +94,23 @@ const getParams = (paramName: string): string | null => {
+    return;
+  }
+
+  window.history.replaceState(null, '', `/?${generateProjectQuery(project)}&${formatParam(queryName, filterParam)}`)
```
I think we need to:
- improve the `generateProjectQuery` so that it doesn't add unchanged/default options
- add a debounce (~ 1 or 2 seconds) to avoid changing it too much 
- avoid adding the `&` in the end when the filter is empty

RubensKj replied:
Cool! great points that you've looked!!

I'll look about it

RubensKj replied:
What I did in the new code..

* Improve `generateProjectQuery` , now it just add what is not default.
* Added a debounce for project updates and filter updates to avoid changing it too much.
* Fixing the query param to avoid the `&` in the end

:)

ia3andy reviewed (commented):
Thanks! I added my review, what do you think?

RubensKj commented:
If you give me the ok about the improve in `generateProjectQuery` I will fix those tests. Or another way I should improve the `generateProjectQuery`.

RubensKj commented:
I had some issues because I have been forgotten to commit the changes 'extensions-picker.spec.tsx.snap', because I changed the component from TextInput to DebouncedTextInput, so I added input delay on snapshot file too (To make tests works). 

Sorry for all this commits :(. However, all works fine right now!! :)

ia3andy commented:
I will have a look tomorrow, I am in PTO today..

On Wed, Feb 10, 2021 at 2:05 AM Rubens Kleinschmidt Júnior <
notifications@github.com> wrote:

> I had some issues because I have been forgotten to commit the changes
> 'extensions-picker.spec.tsx.snap', because I changed the component from
> TextInput to DebouncedTextInput, so I added input delay on snapshot file
> too (To make tests works).
>
> Sorry for all this commits :(. However, all works fine right now!! :)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/quarkusio/code.quarkus.io/pull/422#issuecomment-776355632>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQ664DBYRSW4EB7A5FBBNDS6HLWLANCNFSM4XBTX73A>
> .
>


-- 
Andy Damevin

RubensKj commented:
Have a great PTO today!!

ia3andy reviewed (changes requested):
It looks good beside my comment for the debounce part.. let me know :)

RubensKj commented on src/main/frontend/src/code-quarkus/pickers/extensions-picker.tsx:
```diff
@@ -311,7 +338,7 @@ export const ExtensionsPicker = (props: ExtensionsPickerProps) => {
             fieldId="search-extensions-input"
           >
             <SearchIcon/>
-            <TextInput
```
Fixed!

What I did was remove the useDebounce because it was not affecting the filter from the TextInput, so I used the useRef with the lodash debounce, to debounce the function instead. And another thing is that I put 1000ms (1 second) because I tested and I think it is good time to change the params.

I changed the snap file again in order to rollback the altering of TextInput.

ia3andy commented:
@RubensKj I add a long meeting, really sorry I will review this on monday. Have good WE and thanks :)

RubensKj commented:
@ia3andy oh that's ok!! Have great weekend and rest too!! See you on monday.

ia3andy commented:
Hey @RubensKj I reviewed your PR :)

I created this commit with some changes that I think are relevant:
https://github.com/ia3andy/code.quarkus.io/commit/31b88e3c489396d57a8d065aab04e8eaf70379c4

- Resolve the query params only once
- Directly debounce in the exported `syncParamsInQuery` and use it as is
- Don't need to send the default they are dealt with by the backend (this is even a good improvement)
- Only use `cn` (client name) in the url for download 

WDYT?

RubensKj commented:
I looked to your commit and I really like it. Very good improvements!! 

One thing is that I see a method that is not being used so I put a comment on there, wdyt?

For me it's all ok, awesome improvements!!

ia3andy commented:
ok @RubensKj I've fixed the tests and pushed to your PR, could you squash it all to 1 commit please?

RubensKj commented:
Sure!! Should I create another pull request??

ia3andy commented:
@RubensKj no you need to pull the changes, then rebase to 1 commit then force push:
```
$ git rebase -i upstream/master
< choose squash for all of your commits, except the first one >
```

RubensKj commented:
I think it worked right now in #422 (b569f73)!!

Really sorry for all this, first time doing this :). Tell me if something is wrong.

Thanks for being patience

ia3andy commented:
@RubensKj when you succeed, you should see only 1 commit and not 33 😂 Good luck

ia3andy commented:
https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history

ia3andy commented:
The idea is that you transform a your commit changes to just one.

RubensKj commented:
@ia3andy I did!!! Very happy here!! 😂😂

I tested and it seems all cool and working.

Thanks for sending me the link, really helpful!!

ia3andy reviewed (approved):
Good job 👏 ! Waiting for CI..

ia3andy commented:
@RubensKj if you want to give it a try before prod: https://stage.code.quarkus.io/

ia3andy commented:
Ha, just found a bug in my code :)

RubensKj commented:
> Ha, just found a bug in my code :)

What do you found??

ia3andy commented:
https://github.com/quarkusio/code.quarkus.io/commit/c14984188ed3add33fe298e319b240d5cbff7dfa

RubensKj commented:
Oh, nice!! You already committed.

RubensKj commented:
Tell me when your fix is on Staging, I'll test the feature and another scenarios

ia3andy commented:
@RubensKj it's ON and pretty cool :)

RubensKj commented:
All looking good!! Awesome, good job @ia3andy !! ✌

ia3andy commented:
You did it @RubensKj thanks to you

ia3andy commented:
@RubensKj it's on production: https://code.quarkus.io/?b=GRADLE&s=YjV.49J

Sorry for the delay :)

RubensKj commented:
Awesome!!! I really enjoyed to create this feature!
